src/ILOManager.sol

```diff
@@ -2,253 +2,395 @@
-    function launch(string calldata projectId, address saleToken) external override onlyProjectAdmin(projectId) {
-        require(block.timestamp > _projects[projectId].launchTime, "LT");
+    function launch(
+        string calldata projectId,
+        address token
+    ) external override onlyProjectAdmin(projectId) {
+        Project memory _project = _projects[projectId];
+        require(_project.status == ProjectStatus.INITIALIZED, 'NA');
+        uint160 sqrtPriceX96 = _project.initialPoolPriceX96;
+        PoolAddress.PoolKey memory poolKey = PoolAddress.PoolKey(
+            token,
+            _project.pairToken,
+            _project.fee
+        );
 
-        uint160 sqrtPriceX96 = _projects[projectId].initialPoolPriceX96;
-        PoolAddress.PoolKey memory poolKey = PoolAddress.PoolKey(saleToken, _projects[projectId].raiseToken, _projects[projectId].fee);
-        
         // flip price and tokens
-        if (saleToken > _projects[projectId].raiseToken) {
+        if (token > _project.pairToken) {
             (poolKey.token0, poolKey.token1) = (poolKey.token1, poolKey.token0);
-            sqrtPriceX96 = uint160(2**192 / sqrtPriceX96);
+            sqrtPriceX96 = uint160(2 ** 192 / sqrtPriceX96);
         }
 
-        address uniV3PoolAddress = _initUniV3PoolIfNecessary(poolKey, sqrtPriceX96);
+        address uniV3PoolAddress = _initUniV3PoolIfNecessary(
+            poolKey,
+            sqrtPriceX96
+        );
 
-        address[] memory initializedPools = _initializedILOPools[projectId];
-        require(initializedPools.length > 0, "NP");
-        for (uint256 i = 0; i < initializedPools.length; i++) {
-            IILOPool(initializedPools[i]).launch(projectId, uniV3PoolAddress, poolKey);
+        EnumerableSet.AddressSet
+            storage initializedPools = _initializedILOPools[projectId];
+        uint256 length = EnumerableSet.length(initializedPools);
+        require(length > 0, 'NP');
+        for (uint256 i = 0; i < length; i++) {
+            IILOPoolBase(EnumerableSet.at(initializedPools, i)).launch(
+                uniV3PoolAddress,
+                poolKey,
+                sqrtPriceX96
+            );
         }
         IUniswapV3Pool(uniV3PoolAddress).increaseObservationCardinalityNext(10);
-        emit ProjectLaunch(projectId, uniV3PoolAddress, saleToken);
+        emit ProjectLaunch(projectId, uniV3PoolAddress, token);
     }
-    
-    /// @inheritdoc IILOManager
-    function initProjectFee() external override view returns (uint256) {
-        return _initProjectFee;
+
+    function cancelProject(
+        string calldata projectId
+    ) external override ownerOrProjectAdmin(projectId) onlyInitializedProject(projectId) {
+        _cancelProject(_projects[projectId]);
     }
-    
+
+    function removePool(
+        string calldata projectId,
+        address pool
+    )
+        external
+        override
+        onlyInitializedProject(projectId)
+        onlyProjectAdmin(projectId)
```

Comment:
What if the removed pool is a sale pool? could our flow make sure that the Pool is refundable?

Or to be safer, we could just require the salePool hasn't bee started

Reply:
before pool launch, when cancel a pool will mark the pool is cancelled. This will allow users to claim refund.

Reply:
To simplify the UX, let's block the pool removal once it has started.
It's raising too much tech debt to let it happen.